teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,277 @@
+This problem becomes even greater when new users follow documents such as
+`Get started` which often use same name for task and pipeline. In this environment
+multiple users will step on each other's toes, and produced unintended results.
+
+## Motivation
+
+It is natural for a user to follow ways such as a [Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), ReplicaSet, StatefulSet and also
+Tektoncd/Pipeline taskSpec to have a Pipeline with custom tasks embedded.
+Currently TektonCD/Pipeline supports task specifications to be embedded in
+a pipeline for none-custom-task, but not for custom task. If Tektoncd/Pipeline
+also allows a custom task specification to be embedded in a pipeline specification
+then the behavior will be unified with non custom task and there won't be
+any issues in terms of naming conflict when used by multiple users.
+See issue [tektoncd/pipeline#3682](https://github.com/tektoncd/pipeline/issues/3682)
+
+### Goals
+
+1. Allow custom tasks to be embedded in a pipeline specification,
+2. Custom task specification verification/validation should be handed
+over to custom task controllers, custom task specification must not be
+validated by Tektoncd/Pipeline validation logic.
+3. Custom taskSpec should be submitted as part of the runSpec.
+
+### Non-Goals
+
+Custom task controllers are to be developed by other parties. Custom task
+specification should not be validated by Tektoncd/Pipeline webhooks.
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+- Users can put all the information in one pipelineRun CR.
+- Users don't have to manage the custom task CR. Since custom task CR is
+namespace scope. Currently if there are multiple users in the same namespace, they
+will have conflicts when they are using the same name for their custom task CR.
+- In KFP, we need all the templates and task spec live in each pipeline. Currently, 
+having all the custom task templates living in the Kubernetes namespace scope means that
+we have to make multiple API calls to Kubernetes in order to get all the pipeline
+information to render in our API/UI. 
+
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
+
+- The Tekton controller is responsible for adding the the custom task spec to
+the Run spec. Then it get validate by the custom task controllers.
+
+## Proposal
+
+Use combination of taskRef and taskSpec to indicate an embedded custom task.
```

Comment:
This can go in the alternative section if needed. Combination of `taskRef` and `taskSpec` is generally avoided so far, it's either one or the other allowed. 

This is not needed in the proposal. Just `taskSpec` is sufficient to inline custom specifications. Like @ImJasonH suggested in one of the [comments](https://github.com/tektoncd/pipeline/issues/3682#issuecomment-769519731) in the issue, `taskSpec` can hold `apiVersion`, `kind`, and `spec`. 

This would require to change the `EmbeddedTask` definition to:

```
type EmbeddedTask struct {
	// +optional
	Metadata PipelineTaskMetadata `json:"metadata,omitempty"`

	// TaskSpec is a specification of a task
	TaskSpec `json:",inline,omitempty"`

        // APIVersion and Kind can be specified here.
	runtime.TypeMeta `json:",inline,omitempty"`

        // inline custom task specification here
	Spec             runtime.RawExtension `json:"spec,omitempty"`
}
```

Reply:
With this change to the `EmbeddedTask` struct, inlining `task-loop` specifications can be done with:

```
spec:
  tasks:
    - name: run-my-tests
      taskSpec:
        apiVersion: custom.tekton.dev/v1alpha1
        kind: TaskLoop
        spec:
          taskRef:
            name: simpletask
          iterateParam: word
          timeout: 60s
          retries: 2
```